File: Chap03/Scala-Java/src/main/java/fn/HighOrder.java
package fn;

import java.util.List;
import java.util.stream.IntStream;
import java.util.stream.Collectors;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.function.Consumer;
import java.util.function.BiConsumer;
import java.util.function.IntPredicate;
import java.util.function.IntUnaryOperator;
import java.util.function.IntBinaryOperator;

public class HighOrder {

    public void exec() {

        Consumer<String> pln = System.out::println;
        BiConsumer<String,IntStream> p = (s,a) -> {
            System.out.println(s + a.mapToObj(Integer::valueOf).collect(Collectors.toList()));
        };

        Supplier<IntStream> data = () -> {
            return IntStream.range(0,11);
        };

        pln.accept("---- Java High Order ----");
        p.accept("data->", data.get());

        // functions
        IntPredicate   filter = (a) -> a % 2 == 0;
        IntUnaryOperator  map = (a) -> 2 * a;
        IntBinaryOperator sum = (a,b) -> a + b;

        // argument
        p.accept("filter->", data.get().filter(filter));
        p.accept("filter-> map->", data.get().filter(filter).map(map));
        pln.accept("filter-> map-> reduce->" + data.get().filter(filter).map(map).reduce(sum));

        // currying

        Supplier<Function<IntStream, Integer>> currying = () -> {
            return (a) -> a.filter(filter).map(map).reduce(sum).getAsInt();
        };

        Function<IntStream, Integer> sumFn = currying.get();
        pln.accept("currying->" + sumFn.apply(data.get()));

    }
}

File: Chap03/Scala-Java/src/main/scala/fn/HighOrder.scala
package fn


object HighOrderFunc {

  def filter_fn(a:Int) : Boolean = {
    return a % 2 == 0
  }

  def map_fn(a:Int) : Int = {
    return 2*a
  }

  def sum_fn(a:Int, b:Int) : Int = {
    return a+b
  }

  def currying() : (List[Int]) => Int = {
    def sum(dat:List[Int]) : Int = {
      return dat.filter(filter_fn).map(map_fn).reduce(sum_fn)
    }
    return sum
  }


  def run() {

    println("--- High order Scala ---")

    val data = Range(0,11).toList

    // argument
//    println("data ->" + data)
    println("filter->" + data.filter(filter_fn))
    println("filter-> map->" + data.filter(filter_fn).map(map_fn))
    println("filter-> map-> reduce->" + data.filter(filter_fn).map(map_fn).reduce(sum_fn))

    // currying
    val sumFn = currying()
    println("currying->" + sumFn(data))
  }

}
